src/landlab/components/soil_grading/soil_grading.py

```diff
@@ -0,0 +1,563 @@
+        is_size = np.size(grading_name_split) > 2
+        is_N = len(grading_name_split[0]) == 2 and grading_name_split[0][1].isnumeric()
+        is_numeric = True
+        sum_values = 0
+        for s in grading_name.split("-")[1:]:
+            if s != "spread":
+                is_numeric *= s.isnumeric()
+                if is_numeric:
+                    sum_values += int(s)
+        is_smaller_than_100 = sum_values <= 100 or "spread" in grading_name
+        if is_p * is_size * is_numeric * is_smaller_than_100 * is_N is False:
+            raise ValueError("grading name provided not valid")
+
+        # Create grading distribution based on the grading name
+        N = int(grading_name_split[0][1])
+        parent = np.copy(grain_max_size)
+        limits = np.zeros((n_sizes, 2))
+        limits[0, 1] = grain_max_size
+        for cnt in range(0, n_sizes):
+            limits[cnt, 1] = parent
+            parent = parent * (1 / N) ** (power_of)
+            limits[cnt, 0] = parent
+        limits = limits[::-1]  # sort in ascend order
+        meansizes = np.asarray((limits[:, 0] + limits[:, 1]) / 2)
+
+        # Create fragmentation vector describing the fragmentation pattern
+        precents = np.array(
+            [
+                float(s)
+                for s in grading_name.split("-")
+                if s.replace(".", "", 1).isdigit()
+            ]
+        )
+        if "spread" in grading_name:
+            precents_to_add = np.ones((1, n_sizes)) * precent_of_volume_in_spread
+            precents = np.append(precents, precents_to_add)
+        fragmentation_pattern = precents / 100
+
+        return meansizes, limits, fragmentation_pattern
+
+    def check_fragmentation_pattern(self):
+        # is_length = len(self._fragmentation_pattern) >= 2 * (
+        #     len(self._fragmentation_pattern) <= len(self._meansizes)
+        # )
+        # is_1 = np.sum(self._fragmentation_pattern) <= 1
+        # if not is_length * is_1 is True:
+        #     raise ValueError("fragmentation pattern provided not valid")
+        if (
+            len(self._fragmentation_pattern) < 2
+            or len(self._fragmentation_pattern) > len(self._meansizes)
+            or np.sum(self._fragmentation_pattern) > 1.0
+        ):
+            raise ValueError("fragmentation pattern provided not valid")
+
+    def create_transition_mat(self):
+        """
+        This procedure creates a transition matrix that control the
+        weathering / fragmentation pattern. The matrix represent the
+        proportion of mass that weathered from each size class to other classes
+        """
+
+        self._A = np.zeros((self._n_sizes, self._n_sizes))
+        self._A_factor = np.ones_like(self._A) * self._A_factor
+
+        for i in range(self._n_sizes):
+
+            if i == 0:
+                self._A[i, i] = 0
+            elif i == self._n_sizes:
+                self._A[i, i] = -(1 - (self._fragmentation_pattern[0]))
```

Comment:
@yuvalsmilo This line is never reached. Since the loop runs from 0 to `self._n_sizes - 1`, should this be,
```python
elif i == self._n_sizes - 1:
    self._A[i, i] = -(1 - self._fragmentation_pattern[0])
```

Reply:
There's actually no need to have the *if* statement in the loop at all, is that right?
```python
self._A[0, 0] = 0.0
self._A[self._n_sizes - 1, self._n_sizes - 1] = -(1 - self._fragmentation_pattern[0])

for i in range(1, self._n_sizes - 1):
    self._A[i, i] = -(1 - self._fragmentation_pattern[0])
    ...

```

Reply:
@mcflugen, thanks. I have cleaned this code section.